fpga_interchange/arch.h

```diff
@@ -0,0 +1,1121 @@
+        return id(std::string(site.name.get()) +
+                "/" + IdString(locInfo(bel).bel_data[bel.index].name).str(this));
+    }
+
+    uint32_t getBelChecksum(BelId bel) const { return bel.index; }
+
+    void bindBel(BelId bel, CellInfo *cell, PlaceStrength strength)
+    {
+        NPNR_ASSERT(bel != BelId());
+        NPNR_ASSERT(tileStatus[bel.tile].boundcells[bel.index] == nullptr);
+
+        tileStatus[bel.tile].boundcells[bel.index] = cell;
+        cell->bel = bel;
+        cell->belStrength = strength;
+        refreshUiBel(bel);
+    }
+
+    void unbindBel(BelId bel)
+    {
+        NPNR_ASSERT(bel != BelId());
+        NPNR_ASSERT(tileStatus[bel.tile].boundcells[bel.index] != nullptr);
+        tileStatus[bel.tile].boundcells[bel.index]->bel = BelId();
+        tileStatus[bel.tile].boundcells[bel.index]->belStrength = STRENGTH_NONE;
+        tileStatus[bel.tile].boundcells[bel.index] = nullptr;
+        refreshUiBel(bel);
+    }
+
+    bool checkBelAvail(BelId bel) const
+    {
+        return tileStatus[bel.tile].boundcells[bel.index] == nullptr;
+    }
+
+    CellInfo *getBoundBelCell(BelId bel) const
+    {
+        NPNR_ASSERT(bel != BelId());
+        return tileStatus[bel.tile].boundcells[bel.index];
+    }
+
+    CellInfo *getConflictingBelCell(BelId bel) const
+    {
+        NPNR_ASSERT(bel != BelId());
+        return tileStatus[bel.tile].boundcells[bel.index];
+    }
+
+    BelRange getBels() const
+    {
+        BelRange range;
+        range.b.cursor_tile = 0;
+        range.b.cursor_index = -1;
+        range.b.chip = chip_info;
+        ++range.b; //-1 and then ++ deals with the case of no Bels in the first tile
+        range.e.cursor_tile = chip_info->width * chip_info->height;
+        range.e.cursor_index = 0;
+        range.e.chip = chip_info;
+        return range;
+    }
+
+    Loc getBelLocation(BelId bel) const
+    {
+        NPNR_ASSERT(bel != BelId());
+        Loc loc;
+        getTileXY(bel.tile, &loc.x, &loc.y);
+        loc.z = bel.index;
+        return loc;
+    }
+
+    BelId getBelByLocation(Loc loc) const;
+    BelRange getBelsByTile(int x, int y) const;
+
+    bool getBelGlobalBuf(BelId bel) const
+    {
+        // TODO: This probably needs to be fixed!
+        return false;
+    }
+
+    bool getBelHidden(BelId bel) const {
+        return locInfo(bel).bel_data[bel.index].category != BEL_CATEGORY_LOGIC;
+    }
+
+    IdString getBelType(BelId bel) const
+    {
+        NPNR_ASSERT(bel != BelId());
+        return IdString(locInfo(bel).bel_data[bel.index].type);
+    }
+
+    std::vector<std::pair<IdString, std::string>> getBelAttrs(BelId bel) const;
+
+    int getBelPinIndex(BelId bel, IdString pin) const {
+        NPNR_ASSERT(bel != BelId());
+        int num_bel_wires = locInfo(bel).bel_data[bel.index].num_bel_wires;
+        const int32_t *ports = locInfo(bel).bel_data[bel.index].ports.get();
```

Comment:
There's a marginal performance improvement that some arches take advantage of here by sorting the list of bel ports by IdString and then doing a binary search.

Reply:
I specifically moved from `Array of Struct` to `Struct of Array` because of this point to enable a fast binary search.  However I believe that in most cases, the number of BEL pins is small enough that linear scan is faster than binary search.  Only the large sites would be able to take advantage.

Anyways, for now I'll leave this as a linear scan because the initial BELs used will all be small (e.g. FF's, LUT's, INBUF, OUTBUF).

I think if we are concerned about speeding up the `getBelPinX` functions, there are bigger fish to fry.  Specifically
 - `getBelPins` should be returning a `std::span` (or equiv) instead of a `std::vector`
 - The `IdString pin` methods should be broken up into:
   - `BelPinId getBelPinByName(BelId bel, IdString pin)`
   - `BelId getBelPinBel(BelPinId bel)`
   - `StringId getBelPinName(BelPinId bel)`
   - And change the signature of `getBelPinType`, `getBelPinWire`, and `getBelPins` to:
     - `PortType getBelPinType(BelPinId belpin)`
     - `WireId getBelPinWire(BelPinId belpin)`
     - `const_range<BelPinId> getBelPins(BelId bel)`

For old implementations, `BelPinId` could simply be:
```
typedef BelPin BelPinId;
```
And the new arch methods would just be inlines to the old forms.

I'll make a TODO to revisit this once things are running.

Reply:
That all seems reasonable - `getBelPins` is low priority because it is generally not on any hot paths; whereas `getBelPinWire` is likely to be closer to the hot path (I know it was on the critical path of the placer a long time ago, that's why we did the binary search, but I also suspect that's no longer true after various other changes)

Reply:
If `getBelPinWire` is a hot path, then I believe transitioning to `WireId getBelPinWire(BelPinId belpin)` is likely the right path.  In general, cell <-> BEL pin mapping is something that should come straight from the BBA.  As I start working on the FPGA interchange packer, I believe it will be super clear that something like:
```
const_range<BelPinId> getBelPinsForCellPin(CellInfo *cell_info, IdString s);
```

is the way to go.  I might work on the FPGA interchange packer today to start to get my teeth into those issues.  But I still want to verify that the BBA routing graph is good.

Reply:
This makes sense, although it will need a bit of care implementation-wise to ensure it can deal with things like Xilinx BRAM write enables where the mapping depends on mode (iirc). 

The bigger problem is that the entirety of nextpnr assumes a 1:1 cardinality between cell pins (and consequently sink ports on nets) and bel pins (e.g. timing analysis, placement, routing) - in particular that you can call `getNetinfoSinkWire` for a single port and get a single wire to route to out of it. This is by no means insurmounteable, but it will be a big patch touching some far corners of nextpnr and needing substantial testing (which I do appreciate is a bit of a weak point nextpnr wise currently too).

While we think about this we should also think about fixing the current assumption that cell type and bel type are equivalent; as this potentially involves some similar considerations.

Reply:
The alternative approach would be to have cell pins be physical pins in nextpnr; and have a new unordered_map (possibly in `ArchCellInfo` so it is local to the interchange arch) that correlates back to the logical pin. This would be a lower risk/effort option; at the expense of leaving things a bit less clean.

Reply:
> While we think about this we should also think about fixing the current assumption that cell type and bel type are equivalent; as this potentially involves some similar considerations.

This definitely needs to be decoupled.  The weird thing is that we already have `isValidBelForCell`, so I'm unclear why cell type and BEL type considered equivalent at all?  There is literally an Arch API that answers the question.

For more details, my place for Cell Pin <-> BEL pin mapping per parameters was to apply that in `bindBel`, but maybe I'm missing something.

Reply:
> This definitely needs to be decoupled. The weird thing is that we already have isValidBelForCell, so I'm unclear why cell type and BEL type considered equivalent at all? There is literally an Arch API that answers the question.

In the very early, early days of nextpnr this is how it worked; but over time we needed to partition things in the placer between different cell types and that's why the link was added. I think the entire placement API could be overhauled though - the link between `isValidBelForCell` and `isBelLocationValid` is on the murky side.

Reply:
> For more details, my plan for Cell Pin <-> BEL pin mapping per parameters was to apply that in bindBel, but maybe I'm missing something.

If the mapping is known to be the same for all bel locations doing it `assignArchInfo` will be much faster, as `bindBel` is called a very large number of times during placement legalisation and refinement swaps.

Reply:
> In the very early, early days of nextpnr this is how it worked; but over time we needed to partition things in the placer between different cell types and that's why the link was added

I'd argue that in this case, what is needed is an Arch API that is basically `PartitionId placerPartion(CellInfo *)`, or equiv.  This would break the tight link.  I haven't read the relevant code yet, but this seems fixable without tearing up too much of the world.  But I could be wrong.

Reply:
Agreed - I have this as a special-case for HeAP (the `cellGroups` field in `PlacerHeapCfg`)  already for dealing with things like treating FFs and LUTs as part of the same partition (because you want them to be spread together as they are available in the more-or-less same places). Making this into a generic API would make a lot of sense.

Reply:
I'll look at the code in the placer, but I think we probably need to refactor that placement APIs.  We need a fast "give me available placements for this cell".  This method should be fast, and provide a range of placements.  Maybe this needs a geometric bound?  Unclear.

The other method is narrow (one cell and one BEL) and is allowed to be slower, and that is "the placer believes this is a good place for this Cell, return true if the placement is legal".  One option here is to have "bindBel" return true/false, and only return true if the placement is successful?

Reply:
I'll spend some time reviewing the placer code and make a proposal.

Reply:
> One option here is to have "bindBel" return true/false, and only return true if the placement is successful?

One concern with this - sometimes you might want to swap a lot of bels (for example, swap two whole sites/chains, which isn't something the placer currently does but is on the TODO list) in one go; and then run the legality check after doing the swap. Although incremental checking would improve this, I think you'd still end up with more repeated work doing the checks in `bindBel` rather than at the end of all the binding/unbinding.

Reply:
> The bigger problem is that the entirety of nextpnr assumes a 1:1 cardinality between cell pins (and consequently sink ports on nets) and bel pins (e.g. timing analysis, placement, routing) - in particular that you can call getNetinfoSinkWire for a single port and get a single wire to route to out of it. This is by no means insurmounteable, but it will be a big patch touching some far corners of nextpnr and needing substantial testing (which I do appreciate is a bit of a weak point nextpnr wise currently too).

I know this is future work, but for router based LUT rotation, you likely want a set of candidate sinks for a particular cell port.  This is different than the 1 cell pin to multiple BEL pin case, but looks similar to the router.

I'd argue that getNetinfoSinkWire  should probably return something like:
```
struct NetSinks {
  const_range<WireId> candidate_sink_wires;
  size_t required_sinks;
};
```

E.g. in the LUT rotation case you might return 5 candidate wires, and require 1 or 2 of them to be routed.  In the `WEA`/`WEBWE` case, you would require all of them to be routed.   This idea is only half baked, but it's something to think about for the future.

Reply:
> > One option here is to have "bindBel" return true/false, and only return true if the placement is successful?
> 
> One concern with this - sometimes you might want to swap a lot of bels (for example, swap two whole sites/chains, which isn't something the placer currently does but is on the TODO list) in one go; and then run the legality check after doing the swap. Although incremental checking would improve this, I think you'd still end up with more repeated work doing the checks in `bindBel` rather than at the end of all the binding/unbinding.

Valid concern, I'll think about this.